### PR TITLE
std::ifstream: Defensive Patterns

### PR DESCRIPTION
**Pull request by ax3l**
Add failure handling if inputs in `std::ifstream`s cannot be opened or have problems seek-ing through them.

This should catch I/O errors early.

### CODE REVIEW

**Review by RevathiJambunathan — APPROVED** [review]

**Review by roelof-groenewald — APPROVED** [review]
Thanks!

**Comment by ax3l** [comment]
### Tests that crash with these extra checks

cartesian2d:
- `background_mcc` in `Source/BoundaryConditions/WarpXEvolvePML.cpp:134` (`Source/Diagnostics/Diagnostics.cpp:95`)

cartesian3d:
- `reduced_diags_loadbalancecosts_heuristic` in `Source/BoundaryConditions/WarpX_PEC.cpp:65` (`Source/Diagnostics/ReducedDiags/LoadBalanceCosts.cpp:303`)
- `reduced_diags_loadbalancecosts_timers`

python:
- `Python_background_mcc` in `readCrossSectionFile`

File: Source/Particles/Collision/MCCProcess.cpp [hunk 1]
```diff
@@ -97,6 +97,9 @@ MCCProcess::readCrossSectionFile (
                                   amrex::Gpu::HostVector<amrex::Real>& sigmas )
 {
     std::ifstream infile(cross_section_file);
+    if(!infile) amrex::Abort("Failed to open readCrossSectionFile file");
```
Comment (roelof-groenewald) [inline Source/Particles/Collision/MCCProcess.cpp]: Based on the description in [this](https://gehrcke.de/2011/06/reading-files-in-c-using-ifstream-dealing-correctly-with-badbit-failbit-eofbit-and-perror/) article should we maybe change this whole `readCrossSectionFile` function to:
```
    std::ifstream infile(cross_section_file);
    if(!infile.is_open()) amrex::Abort("Failed to open cross-section data file");

    double energy, sigma;
    while (infile >> energy >> sigma) {
        energies.push_back(energy);
        sigmas.push_back(sigma);
    }
    if (infile.bad()) amrex::Abort("Failed to read cross-section data from file.");
    infile.close();
```

Comment (ax3l) [reply Source/Particles/Collision/MCCProcess.cpp]: That looks good to me as well :)

**Pull request merged**